### PR TITLE
feat: Checking socks5 proxy before launching a scan #1001

### PR DESCRIPTION
**Pull request by LuitelSamikshya**
## Checklist

- [x] Pull request is created against the [dev](https://github.com/projectdiscovery/nuclei/tree/dev) branch
- [x] All checks passed (lint, unit/integration/regression tests etc.) with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: v2/internal/runner/options.go [hunk 1]
```diff
@@ -167,6 +173,101 @@ func loadResolvers(options *types.Options) {
...
+		if options.ProxyURL == "" && options.ProxySocksURL == "" {
+			if valid := assignProxy(proxy, options); valid {
+				done <- true
+			}
+		}
+	}
+	exitCounter <- true
+}
+func testProxyConnection(proxy string, timeoutDelay int) (string, error) {
+	ip, _ := url.Parse(proxy)
```
[the remark this answers is not in the document]
Comment (LuitelSamikshya) [reply v2/internal/runner/options.go]: It's taken care of in the latest commit.

File: v2/pkg/types/types.go [hunk 2]
```diff
@@ -49,10 +51,14 @@ type Options struct {
 	TargetsFilePath string
 	// Output is the file to write found results to.
 	Output string
+	// List of proxy server (file or comma seperated)
+	Proxy string
```
[the remark this answers is not in the document]
Comment (LuitelSamikshya) [reply v2/pkg/types/types.go]: Done

File: README.md [hunk 3]
```diff
@@ -162,9 +162,8 @@ DEBUG:
    -debug                     show all requests and responses
    -debug-req                 show all sent requests
    -debug-resp                show all received responses
-   -proxy, -proxy-url string  URL of the HTTP proxy server
-   -proxy-socks-url string    URL of the SOCKS proxy server
-   -tlog, -trace-log string   file to write sent requests trace log
+   -proxy string              list of proxy server (file or comma separated)
```
[the remark this answers is not in the document]
Comment (LuitelSamikshya) [reply README.md]: Done

File: v2/cmd/nuclei/main.go [hunk 4]
```diff
@@ -134,11 +134,9 @@ on extensive configurability, massive extensibility and ease of use.`)
 		flagSet.BoolVar(&options.DebugRequests, "debug-req", false, "show all sent requests"),
 		flagSet.BoolVar(&options.DebugResponse, "debug-resp", false, "show all received responses"),
 
-		/* TODO why the separation? http://proxy:port vs socks5://proxy:port etc
-		   TODO should auto-set the HTTP_PROXY variable for the process? */
-		flagSet.StringVarP(&options.ProxyURL, "proxy-url", "proxy", "", "URL of the HTTP proxy server"),
-		flagSet.StringVar(&options.ProxySocksURL, "proxy-socks-url", "", "URL of the SOCKS proxy server"),
-		flagSet.StringVarP(&options.TraceLogFile, "trace-log", "tlog", "", "file to write sent requests trace log"),
+		/* TODO should auto-set the HTTP_PROXY variable for the process? */
+		flagSet.StringVar(&options.Proxy, "proxy", "", "list of proxy server (file or comma separated)"),
```
[the remark this answers is not in the document]
Comment (LuitelSamikshya) [reply v2/cmd/nuclei/main.go]: Done

File: v2/internal/runner/proxy.go [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+var proxyURLList []url.URL
+
+// loadProxyServers load list of proxy servers from file or comma seperated
+func loadProxyServers(options *types.Options) error {
+	if len(options.Proxy) == 0 {
+		return nil
+	}
+	for _, p := range options.Proxy {
+		if strings.TrimSpace(p) == "" {
```
Comment (forgedhallpass) [inline v2/internal/runner/proxy.go:26]: NormalizedStringSlices should not contain empty/blank elements by design

File: v2/internal/runner/proxy.go [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+					}
+				}
+			}
+		}
+	}
+}
+
+func runProxyConnectivity(proxyURL url.URL, options *types.Options, done chan bool, exitCounter chan bool) {
+	if err := testProxyConnection(proxyURL, options.Timeout); err == nil {
+		// if options.ProxyURL == "" && options.ProxySocksURL == "" {
```
Comment (forgedhallpass) [inline v2/internal/runner/proxy.go:90]: I did not mean to replace the original logic with the environmental variable, but to add it as an extension to it. There were/are cases where the original proxy logic was not applicable, hence adding the environmental variable option would handle those cases.

Comment (forgedhallpass) [reply v2/internal/runner/proxy.go:90]: This should only be done after you found a valid/working proxy.

File: v2/internal/runner/proxy.go [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+	if url, err := url.Parse(proxy); err == nil && isSupportedProtocol(url.Scheme, false) {
+		return *url, nil
+	}
+	return url.URL{}, errors.New("invalid proxy format (It should be http[s]/socks5://[username:password@]host:port)")
+}
+
+//isSupportedProtocol checks given protocols are supported
+func isSupportedProtocol(value string, prefixCheck bool) bool {
+	if prefixCheck {
+		value = strings.ToLower(value)
```
Comment (forgedhallpass) [inline v2/internal/runner/proxy.go:141]: The `url.Scheme` should already contain the value lower cased. I'd change the method parameter to `url` and then you can remove the `ToLower` part

Comment (LuitelSamikshya) [reply v2/internal/runner/proxy.go]: Done

File: v2/internal/runner/proxy.go (hunk 6 above)
Comment (LuitelSamikshya) [reply v2/internal/runner/proxy.go]: Done

File: v2/internal/runner/proxy.go (hunk 5 above)
Comment (LuitelSamikshya) [reply v2/internal/runner/proxy.go]: Done

File: v2/cmd/nuclei/main.go [hunk 8]
```diff
@@ -134,11 +134,9 @@ on extensive configurability, massive extensibility and ease of use.`)
 		flagSet.BoolVar(&options.DebugRequests, "debug-req", false, "show all sent requests"),
 		flagSet.BoolVar(&options.DebugResponse, "debug-resp", false, "show all received responses"),
 
-		/* TODO why the separation? http://proxy:port vs socks5://proxy:port etc
-		   TODO should auto-set the HTTP_PROXY variable for the process? */
-		flagSet.StringVarP(&options.ProxyURL, "proxy-url", "proxy", "", "URL of the HTTP proxy server"),
-		flagSet.StringVar(&options.ProxySocksURL, "proxy-socks-url", "", "URL of the SOCKS proxy server"),
-		flagSet.StringVarP(&options.TraceLogFile, "trace-log", "tlog", "", "file to write sent requests trace log"),
+		/* TODO should auto-set the HTTP_PROXY variable for the process? */
```
[the remark this answers is not in the document]
Comment (LuitelSamikshya) [reply v2/cmd/nuclei/main.go:137]: Done

[the remark this answers is not in the document]
Comment (forgedhallpass) [reply v2/cmd/nuclei/main.go:137]: please also remove the TODO

File: v2/pkg/protocols/http/httpclientpool/clientpool.go [hunk 9]
```diff
@@ -172,27 +171,32 @@ func wrappedGet(options *types.Options, configuration *Configuration) (*retryabl
...
-		if proxyErr == nil {
-			transport.DialContext = dc.DialContext
-		}
-	}
 	if proxyURL != nil {
-		transport.Proxy = http.ProxyURL(proxyURL)
+		// Attempts to overwrite the dial function with the socks proxied version
+		if proxyURL.Scheme == types.SOCKS5 {
+			var proxyAuth *proxy.Auth = &proxy.Auth{}
+			//var proxyAuth *proxy.Auth
```
Comment (forgedhallpass) [inline v2/pkg/protocols/http/httpclientpool/clientpool.go:178]: remove the commented out code if it is not needed anymore

**Review by forgedhallpass — APPROVED** [review]
The PR looks much cleaner now 👍 The PR should be squashed then merged to not pollute the history.

File: v2/pkg/protocols/http/httpclientpool/clientpool.go (hunk 9 above)
Comment (LuitelSamikshya) [reply v2/pkg/protocols/http/httpclientpool/clientpool.go]: Done

**Pull request merged**